File: backend/app/models/message.py
from sqlalchemy import Column, Integer, String, DateTime, ForeignKey, Text, JSON
from sqlalchemy.orm import relationship
from sqlalchemy.sql import func
import enum

from app.database import Base


class MessageType(str, enum.Enum):
    TEXT = "text"
    SYSTEM = "system"
    CHART = "chart"  # 차트 공유 메시지


class Message(Base):
    __tablename__ = "messages"

    id = Column(Integer, primary_key=True, index=True)
    discussion_id = Column(Integer, ForeignKey("discussions.id", ondelete="CASCADE"), nullable=False, index=True)
    user_id = Column(Integer, ForeignKey("users.id"), nullable=True)

    content = Column(Text, nullable=False)
    message_type = Column(String(20), default=MessageType.TEXT.value)
    chart_data = Column(JSON, nullable=True)  # 차트 캔들 데이터 (message_type='chart' 일 때)

    created_at = Column(DateTime(timezone=True), server_default=func.now(), index=True)

    # Relationships
    discussion = relationship("Discussion", back_populates="messages")
    user = relationship("User", back_populates="messages")

File: spring-backend/backend/src/main/java/com/fundmessenger/discussion/entity/Message.java
package com.fundmessenger.discussion.entity;

import com.fundmessenger.user.entity.User;
import io.hypersistence.utils.hibernate.type.json.JsonType;
import jakarta.persistence.*;
import lombok.AllArgsConstructor;
import lombok.Getter;
import lombok.NoArgsConstructor;
import lombok.Setter;
import org.hibernate.annotations.CreationTimestamp;
import org.hibernate.annotations.OnDelete;
import org.hibernate.annotations.OnDeleteAction;
import org.hibernate.annotations.Type;

import java.time.OffsetDateTime;

@Entity
@Table(name = "messages")
@Getter
@Setter
@NoArgsConstructor
@AllArgsConstructor
public class Message {

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    @Column(name = "id")
    private Long id;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "discussion_id", nullable = false)
    @OnDelete(action = OnDeleteAction.CASCADE)
    private Discussion discussion;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "user_id")
    private User user;

    @Column(name = "content", columnDefinition = "text", nullable = false)
    private String content;

    @Column(name = "message_type", length = 20, columnDefinition = "varchar(20) default 'text'")
    private String messageType = "text";

    @Type(JsonType.class)
    @Column(name = "chart_data", columnDefinition = "jsonb")
    private Object chartData;

    @Column(name = "session_number", columnDefinition = "integer default 1")
    private Integer sessionNumber = 1;

    @CreationTimestamp
    @Column(name = "created_at", columnDefinition = "timestamptz")
    private OffsetDateTime createdAt;
}
